### DISCUSSION TITLE
Arduino Uno R4 Wifi Compatibility

### DISCUSSION THREAD

**Discussion by tomasludrovan** [Makuna/NeoPixelBus#866] [General]
Hi all & special thanks to the author of this great library!

I am having difficulties with running the library on the Arduino Uno R4 Wifi - it seems to be not supported yet. It is clashing with ARM architecture I guess.

Is there any possibility/workaround to run it on this board?

Thank you!

**Comment by Makuna** [comment]
When you compile with verbose output on, you will find a line in the output that is very, very long that will contain `ARDUINO_ARCH_`* and/or `__arm__`, usually proceeded with a `-D`. 
Copy that line and reply with it.

ARM is not pleasant to support.  They are not as compatible as the "ARM" category would imply.  All the current support is BitBang, meaning there is no hardware support to output the bits so that to maintain the timing, all ISRs are suspended while sending the data.  This is no different than AVR, but they don't have WiFi.

**Reply by tomasludrovan** [reply]
Hi,

thanks for your quick reply. I understand that it is tricky to support multiple CPU architectures.

**Precondition:**

- Board Manager - install Arduino UNO R4 Boards

**Here is the verbose output of the compiler:**
C:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4/bin/arm-none-eabi-g++ -c -w -Os -g3 -fno-use-cxa-atexit -fno-rtti -fno-exceptions -nostdlib -DF_CPU=48000000 -DNO_USB -DBACKTRACE_SUPPORT -DARDUINO_UNOR4_WIFI -std=gnu++17 -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsigned-char -ffunction-sections -fdata-sections -fmessage-length=0 -fno-builtin -w -x c++ -E -CC -DARDUINO=10607 -DPROJECT_NAME="C:\Users\Tomas\AppData\Local\Temp\arduino\sketches\465A66AD941963E465F2FF198DAD8C9D/LedController.ino" -DARDUINO_UNOWIFIR4 -DARDUINO_ARCH_RENESAS_UNO -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 -mthumb @C:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/defines.txt -DCFG_TUSB_MCU=OPT_MCU_RAXXX -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/tinyusb -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated-avr-comp -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4 -IC:\Users\Tomas\Documents\Arduino\libraries\NeoPixelBus_by_Makuna\src -IC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\libraries\SPI -iprefixC:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2 @C:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/includes.txt C:\Users\Tomas\AppData\Local\Arduino15\packages\arduino\hardware\renesas_uno\1.3.2\libraries\SPI\SPI.cpp -o nul


**The first error shown is:**
NeoArmMethod.h:763:26: error: 'VARIANT_MCK' was not declared in this scope
 #define ARM_OTHER_SCALE  VARIANT_MCK / 2UL / 1000000UL

**Comment by Makuna** [comment]
ARDUINO_ARCH_RENESAS
ARDUINO_ARCH_RENESAS_UNO,

but there are some hidden ones not there, probably hidden in the "variants\UNOWIFIR4/defines.txt".

Could you pick up the Master version and give it a try.  I just merged in a branch that I was working on for improved ARM support.

**Reply by tomasludrovan** [reply]
Hi Michael,

I tested with master - same result:
NeoArmMethod.h:777:26: error: 'VARIANT_MCK' was not declared in this scope
 #define ARM_OTHER_SCALE  VARIANT_MCK / 2UL / 1000000UL

What do you think about tuning the VARIANT_MCK parameter? Or do you have any other ideas, which I can try?

Here is the content of defines.txt for UNOWIFIR4:
-D_RA_CORE=CM4
-D_RENESAS_RA_

**Comment by acarito** [comment]
Can I #define this constant in my code before including the library on the Arduino R4 and use the bigbang method? Would I set it at the clock speed or something else?
